sw/host/opentitansession/src/main.rs

```diff
@@ -96,6 +113,119 @@ fn parse_command_line(opts: Opts, mut args: ArgsOs) -> Result<Opts> {
+            Ok(Box::new(result))
+        }
+        Ok(Err(e)) => bail!(e),
+        Err(e) => bail!("Child process failed to start: {}", e),
+    }
+}
+
+// This method runs in the daemon child.  It will instantiate SessionHandler to bind to a
+// socket, then report the chosen port number to the parent process by means of a serialized
+// `SessionStartResult` sent through the stdout anonymous pipe, and finally enter an infnite
+// loop, processing connections on that socket
+fn session_child(listen_port: Option<u16>, backend_opts: &backend::BackendOpts) -> Result<()> {
+    let transport = backend::create(backend_opts, "null")?;
+    let mut session = SessionHandler::init(&transport, listen_port)?;
+    // Instantiation of Transport backen, and binding to a socket was successful, now go
+    // thought the process of making this process a daemon, disconeected from the
+    // terminal that was used to start it.
+
+    env::set_current_dir(&Path::new("/"))?;
+
+    // Close stdin and stderr, which would likely be connected to the tty (terminal)
+    // used to run opentitansession.
+    {
+        let devnull_file = File::open("/dev/null")?;
+        // devnull_fd is valid only as long as devnull_file remains in scope.
+        let devnull_fd: RawFd = devnull_file.as_raw_fd();
+        dup2(devnull_fd, 0)?;
+        dup2(devnull_fd, 2)?;
+    }
+    // After severing any stdio/err connections to the controlling terminal of the
+    // parent, we can now establish new Unix "session" for this process, which will not
+    // be connected to any terminal.  This means that this daemon will not be killed by
+    // SIGHUP, in case the terminal that was used for running `session start` is later
+    // closed.
+    setsid()?;
+
+    // Report startup success to parent process.
+    serde_json::to_writer::<io::Stdout, Result<SessionStartResult, String>>(
+        io::stdout(),
+        &Ok(SessionStartResult {
+            port: session.get_port(),
+        }),
+    )?;
+    io::stdout().flush()?;
+
+    // Closing the standard output pipe is the signal to the parent process that this
+    // child has started up successfully.  We close the pipe indirectly, by replacing
+    // file descriptor 1 with one pointing to /dev/null.
+    dup2(0, 1)?;
+
+    // Indefinitely run command processing loop in this daemon process.
+    session.run_loop()
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct SessionStopResult {}
+
+/// Load .pid file based on given port number, and send SIGTERM to the process identified in the
+/// file, to request the daemon gracefully shut down.
+fn stop_session(run_file_fn: impl FnOnce(u16) -> PathBuf, port: u16) -> Result<Box<dyn Serialize>> {
+    // Read the pid file corresponding to the requested TCP port.
+    let path = run_file_fn(port);
+    let reader = BufReader::new(File::open(&path)?);
+    let pid: i32 = FromStr::from_str(
```

Comment:
You can just use `fs::read_to_string` for this.

Reply:
Thanks for the suggestion, this is simpler.